ako-infra/ingestion/avi_infra_controller.go

```diff
@@ -102,165 +106,231 @@ func (a *AviControllerInfra) DeriveCloudNameAndSEGroupTmpl(tz string) (error, st
-				return false
-			}
-
-		} else {
-			utils.AviLog.Errorf("Get uri %v returned err %v", uri, err)
-			return false
-		}
+		utils.AviLog.Errorf("Get uri %v returned err %v", uri, err)
+		return err, nil
 	}
-	// Construct an SE group based on parameters in the `Default-Group`
+
 	elems := make([]json.RawMessage, result.Count)
 	err = json.Unmarshal(result.Results, &elems)
 	if err != nil {
 		utils.AviLog.Errorf("Failed to unmarshal data, err: %v", err)
-		return false
+		return err, nil
 	}
 
+	// Using clusterID for advl4.
+	clusterName := lib.GetClusterID()
 	for _, elem := range elems {
 		seg := models.ServiceEngineGroup{}
-		if err := json.Unmarshal(elem, &seg); err != nil {
+		err = json.Unmarshal(elem, &seg)
+		if err != nil {
 			utils.AviLog.Warnf("Failed to unmarshal data, err: %v", err)
 			continue
 		}
 
-		if !ConfigureSeGroup(a.AviRestClients.AviClient[0], &seg) {
-			return false
+		if *seg.Name == clusterName {
+			return nil, &seg
 		}
 	}
-	return true
-}
-
-// ConfigureSeGroup creates the SE group with the supplied properties, alters just the SE group name and the labels.
-func ConfigureSeGroup(client *clients.AviClient, seGroup *models.ServiceEngineGroup) bool {
-	// Change the name of the SE group
-	*seGroup.Name = lib.GetClusterID()
-
-	uri := "/api/serviceenginegroup/"
-	// Add the labels.
-	seGroup.Labels = lib.GetLabels()
-	response := models.ServiceEngineGroupAPIResponse{}
-	SetAdminTenant := session.SetTenant(lib.GetAdminTenant())
-	SetTenant := session.SetTenant(lib.GetTenant())
-
-	err := lib.AviPost(client, uri, seGroup, response)
-	if err != nil {
-		if aviError, ok := err.(session.AviError); ok && aviError.HttpStatusCode == 403 {
-			//SE in provider context
-			utils.AviLog.Debugf("Switching to admin context from  %s", lib.GetTenant())
-			SetAdminTenant(client.AviSession)
-			defer SetTenant(client.AviSession)
-			err := lib.AviPost(client, uri, seGroup, response)
-			if err != nil {
-				utils.AviLog.Warnf("Error during POST call to create the SE group :%v", err.Error())
-				return false
 
-			}
-		} else {
-			utils.AviLog.Warnf("Error during POST call to create the SE group :%v", err.Error())
-			return false
+	if result.Next != "" {
+		// It has a next page, let's recursively call the same method.
+		next_uri := strings.Split(result.Next, "/api/serviceenginegroup")
+		if len(next_uri) > 1 {
+			overrideUri := "/api/serviceenginegroup" + next_uri[1]
+			nextPage := lib.NextPage{NextURI: overrideUri}
+			return fetchSEGroup(client, nextPage)
 		}
 	}
-	utils.AviLog.Infof("labels: %v set on Service Engine Group :%v", utils.Stringify(lib.GetLabels()), *seGroup.Name)
-	return true
 
+	utils.AviLog.Infof("No Service Engine Group found for Cluster.")
+	return nil, nil
 }
 
-func checkSeGroup(client *clients.AviClient, cloudName string) bool {
-	segroupName := lib.GetClusterID()
+func fetchVcenterServer(client *clients.AviClient) (string, error) {
+	uri := "/api/vcenterserver"
```

Comment:
with what user infra container will be running? Do we need to check role changes for it?

Reply:
nsxt-alb user, it has the required roles to make this API call